### PR TITLE
FoundationEssentials: improve Windows coverage for `Platform.swift`

### PR DESCRIPTION
**Pull request by compnerd**
Adjust the codepaths to improve Windows support in the file. This is simply blindly trying to match semantics to reduce the errors reported during building `FoundationEssentials`.

### CODE REVIEW

File: Sources/FoundationEssentials/Platform.swift [hunk 1]
```diff
@@ -220,6 +246,15 @@ extension Platform {
...
         return try? FileManager.default.destinationOfSymbolicLink(
             atPath: "/proc/self/exe").standardizingPath
+#elseif os(Windows)
+        let hFile = GetModuleHandleW(nil)
+        let dwLength: DWORD = GetFinalPathNameByHandleW(hFile, nil, 0, FILE_NAME_NORMALIZED)
+        return withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwLength)) { lpBuffer in
+            guard GetFinalPathNameByHandleW(hFile, lpBuffer.baseAddress, dwLength, FILE_NAME_NORMALIZED) == dwLength - 1 else {
+                return nil
+            }
+            return String(decodingCString: lpBuffer.baseAddress!, as: UTF16.self)
```
Comment (jmschonfeld) [inline Sources/FoundationEssentials/Platform.swift:256]: The other platforms call `.standardizingPath` on the result, should we do the same here for consistency?

**Comment by compnerd** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/Platform.swift (hunk 1 above)
Comment (compnerd) [reply Sources/FoundationEssentials/Platform.swift:256]: Technically that is already done - `GetFinalPathNameByHandleW` will resolve the symlink and standardize the path in the process. We get the fully normalized location that the file is at.

Comment (jmschonfeld) [reply Sources/FoundationEssentials/Platform.swift:256]: Ah I see, ok if it's already fully standardized the same way that `standardizePath` works then that seems fine to leave it as-is here. Could we add a comment to note that just so it's clear why it's different than the other OS code paths?

**Comment by jmschonfeld** [comment]
It looks like the `CInt(Platform.MAX_HOSTNAME_LENGTH)` change broke the linux/macOS build:

```
/build/swift-foundation/Sources/FoundationEssentials/Platform.swift:206:48: error: cannot convert value of type 'CInt' (aka 'Int32') to expected argument type 'Int'
204 |     static func getHostname() -> String {
205 |         return withUnsafeTemporaryAllocation(of: CChar.self, capacity: Platform.MAX_HOSTNAME_LENGTH + 1) {
206 |             guard gethostname($0.baseAddress!, CInt(Platform.MAX_HOSTNAME_LENGTH)) == 0 else {
    |                                                `- error: cannot convert value of type 'CInt' (aka 'Int32') to expected argument type 'Int'
207 |                 return ""
208 |             }
```

File: Sources/FoundationEssentials/Platform.swift (hunk 1 above)
Comment (compnerd) [reply Sources/FoundationEssentials/Platform.swift:256]: Sure! That sounds reasonable.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Review by jmschonfeld — APPROVED** [review]
Looks great now, thanks! (and TIL about `numericCast(x)` 🙂)

**Pull request merged**